pkg/gather/clusterconfig/version.go

```diff
@@ -2,75 +2,110 @@ package clusterconfig
+// Location of cluster ID: config/id/
+// Location of pods in archive: config/pod/
+// Location of events in archive: events/
 // Id in config: version
 func GatherClusterVersion(g *Gatherer, c chan<- gatherResult) {
 	defer close(c)
-	config, err := GetClusterVersion(g.ctx, g.gatherKubeConfig)
+	gatherConfigClient, err := configv1client.NewForConfig(g.gatherKubeConfig)
 	if err != nil {
 		c <- gatherResult{nil, []error{err}}
 		return
 	}
-	c <- gatherResult{[]record.Record{{Name: "config/version", Item: record.JSONMarshaller{Object: anonymizeClusterVersion(config)}}}, nil}
-}
-
-func GetClusterVersion(ctx context.Context, kubeConfig *rest.Config) (*configv1.ClusterVersion, error) {
-	gatherConfigClient, err := configv1client.NewForConfig(kubeConfig)
+	gatherKubeClient, err := kubernetes.NewForConfig(g.gatherProtoKubeConfig)
 	if err != nil {
-		return nil, err
+		c <- gatherResult{nil, []error{err}}
+		return
 	}
-	config, err := gatherConfigClient.ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
+	records, errors := getClusterVersion(g.ctx, gatherConfigClient, gatherKubeClient.CoreV1())
+	c <- gatherResult{records, errors}
+}
+
+func getClusterVersion(ctx context.Context, configClient configv1client.ConfigV1Interface, coreClient corev1client.CoreV1Interface) ([]record.Record, []error) {
+	config, err := configClient.ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
 	if errors.IsNotFound(err) {
 		return nil, nil
 	}
 	if err != nil {
-		return nil, err
+		return nil, []error{err}
 	}
-	return config, nil
-}
 
-// GatherClusterID stores ClusterID from ClusterVersion version
-// This method uses data already collected by Get ClusterVersion. In particular field .Spec.ClusterID
-// The Kubernetes api https://github.com/openshift/client-go/blob/master/config/clientset/versioned/typed/config/v1/clusterversion.go#L50
-// Response see https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L38
-//
-// Location in archive: config/id/
-// See: docs/insights-archive-sample/config/id
-// Id in config: id
-func GatherClusterID(g *Gatherer, c chan<- gatherResult) {
-	defer close(c)
-	version, err := GetClusterVersion(g.ctx, g.gatherKubeConfig)
+	records := []record.Record{
+		{Name: "config/version", Item: record.JSONMarshaller{Object: anonymizeClusterVersion(config)}},
+	}
+
+	if config.Spec.ClusterID != "" {
+		records = append(records, record.Record{Name: "config/id", Item: marshal.Raw{Str: string(config.Spec.ClusterID)}})
+	}
+
+	// TODO: In the future, make this conditional on sad ClusterVersion conditions or ClusterVersionOperatorDown alerting, etc.
+	namespace := "openshift-cluster-version"
+	now := time.Now()
+	unhealthyPods := []*corev1.Pod{}
+
+	pods, err := coreClient.Pods(namespace).List(ctx, metav1.ListOptions{})
 	if err != nil {
-		c <- gatherResult{nil, []error{err}}
-		return
+		klog.V(2).Infof("Unable to find pods in namespace %s for cluster-version operator", namespace)
+		return records, nil
```

Comment:
Maybe return the error here as well? Btw what is the expected number of pods in this namespace please? only one?

Reply:
yeah, in most cases there will only be one.  Returning `nil` here is pattern-matching [this existing error-swallowing][1].  If you want me including this error in the return, do you want me to also patch to return that one?

[1]: https://github.com/openshift/insights-operator/blob/7203a31b3e9744b6cafb397ae49a3eab584ee878/pkg/gather/clusterconfig/operators.go#L143-L144

Reply:
Ahh ok. Well it's true that it's not so critical for the gatherer, so we can swallow it.